src/backend/commands/matview.c

```diff
@@ -1055,7 +1058,24 @@ IVM_immediate_maintenance(PG_FUNCTION_ARGS)
 	rel = trigdata->tg_relation;
 	relid = rel->rd_id;
 
-	query = get_view_query(matviewRel);
+	/* get view query*/
+	rule = matviewRel->rd_rules->rules[0];
+	if (rule->event != CMD_SELECT || !(rule->isInstead))
+		elog(ERROR,
+			 "the rule for materialized view \"%s\" is not a SELECT INSTEAD OF rule",
+			 RelationGetRelationName(matviewRel));
+
+	actions = rule->actions;
+	if (list_length(actions) != 1)
+		elog(ERROR,
+			 "the rule for materialized view \"%s\" is not a single action",
+			 RelationGetRelationName(matviewRel));
+
+	/*
+	 * The stored query was rewritten at the time of the MV definition, but
+	 * has not been scribbled on by the planner.
+	 */
+	query = linitial_node(Query, actions);
```

Comment:
Thanks!

I originally consider to fix the Assert condition in get_view_query(), but your solution that doesn't use get_view query and get it by itself seems better because  ExecRefreshMatView() does in this manner!

By the way, is there any reason or intention to check internal errors of only two condition?  ExecRefreshMatView() checks other conditions, for example ""materialized view \"%s\" is missing rewrite information".  If there is no problem, how about making a function to get a view definition query for matviews and calling this from both functions?

Reply:
I think that other check internal errors are unnecessary.
So no problem even if add. I make its function which get a view definition query for matview.

Reply:
I am just interested in why you think some check is necessary and other is not. If there is good reason, I think it is nice to merge this fix. We can make the common function after we recognize making this is needed of better.

Reply:
sorry... I misunderstood. others checks are necessary too.